### PR TITLE
Add support for physical keyboard raw input for iOS

### PR DESCRIPTION
**Pull request by imxieyi**
Need to change [UIApplication entry](https://github.com/swoolcock/osu/blob/samah-ios/osu.iOS/Application.cs#L12) to:
```C#
UIApplication.Main(args, "GameUIApplication", "AppDelegate");
```

---

This PR adds support for physical keyboard (e.g. Smart Keyboard for iPad Pro) on iOS. It captures raw input, i.e. key down & key up, of each single key. It is useful especially for osu!mania mode.

~However, it will hook all key inputs including those from on-screen keyboards.~ A simple solution is to call `handleKeyUIEvent` in base class using `objc_SendMsgSuper`, ~which is not available in C#, in other words, requires a objc plugin.~ This has been solved by importing `objc_SendMsgSuper` to C#.

### CODE REVIEW

**Comment by swoolcock** [comment]
osu!framework already supports the Smart Keyboard, it's how I've been doing testing.  Your solution to catching the keyup event is similar to this post I read recently:
https://nacho4d-nacho4d.blogspot.com/2012/01/catching-keyboard-events-in-ios.html

I'm hesitant to add this to the framework in case Apple turn around and start removing people's apps from the store for fudging about in private code.

File: osu.Framework.iOS/GameUIApplication.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using osu.Framework.Logging;
+using osu.Framework.Input;
+using System;
+
+namespace osu.Framework.iOS {
+
+    [Register("GameUIApplication")]
+    public class GameUIApplication : UIApplication {
+
+        private static bool IS_IOS7 = int.Parse(UIDevice.CurrentDevice.SystemVersion.Split('.')[0]) >= 7;
```
Comment (swoolcock) [inline osu.Framework.iOS/GameUIApplication.cs]: This is not the correct way to check system versions in iOS.

**Comment by swoolcock** [comment]
Additionally, this should really be in osuTK.

**Comment by imxieyi** [comment]
> Please give an example of this method in a current production app that has been accepted by Apple, so we can see if it's feasible to include.

Tried a whole bunch of remote desktop apps (Duet, TeamViewer, RDP, NoMachine, Vnc Viewer). Only [VNC Viewer](https://www.realvnc.com/en/connect/download/viewer/ios/) seems to support raw keyboard input. At least in [online keyboard test](http://en.key-test.ru/) it responds to key up & down for modifier keys (ctrl, shift, option etc.). It is strange that it cannot capture key up & down for letters. Don't know if this apps captures such events using that private api.

**Comment by swoolcock** [comment]
Please create a PR in the osuTK project to fire key events on the `GameWindow` rather than implementing an `InputHandler` here, as it is a fairly low-level change.

**Comment by imxieyi** [comment]
> Please create a PR in the osuTK project to fire key events on the `GameWindow` rather than implementing an `InputHandler` here, as it is a fairly low-level change.

Having been messing around with osuTK for several days, still haven't got it working. Should this be implemented in `osuTK.iOS`? I tried to create a custom `IKeyboardDriver2` instance under `iOSFactory.CreateKeyboardDriver` but this method was never called. Also, it seems that it is not feasible to call `KeyDown` and `KeyUp` events directly in `iOSGameView` due to internal access level of `KeyboardKeyEventArgs`.

**Comment by swoolcock** [comment]
@imxieyi What's the state of this?  Would be nice to get it in so we can start testing game modes other than osu!standard.

**Comment by imxieyi** [comment]
The keyboard input already works naturally now. The modifiers also works in osu!catch mode. There are two problems that I cannot fix for now:
1. The keymap only contains keys on iPad Pro Smart Keyboard. I don't have access to other keyboards yet.
2. The arrow keys will conflict with existing key input based on TextField, which means, these keys will be handled twice in the song selection scene.

**Comment by peppy** [comment]
point 2 sounds like blocking behaviour which we will need to resolve.

**Comment by swoolcock** [comment]
You can probably completely replace the existing `IOSKeyboardHandler` with this implementation, which should solve point 2.

**Comment by imxieyi** [comment]
> You can probably completely replace the existing IOSKeyboardHandler with this implementation, which should solve point 2. At the very least, removing the UIKeyCommands for the arrow keys will help.

I think the best practice is to only let `gsEvent` pass through after `RawHandler` fails. (Currently it let all events go through) Completely replacing existing `IOSKeyboardHandler` would make on-screen keyboard stop working. I will try this method after Chinese New Year.

**Comment by imxieyi** [comment]
Finally decided on blocking arrow keys only. It would be pretty tricky to handle text input in raw handler, especially when multi-language support is needed.

**Comment by swoolcock** [comment]
Just need to make sure that whatever is changed, the default iOS stuff should still work like autocomplete and long press to insert accented characters, etc.  The current keyboard handler is backed by a `UITextField`, so we get this functionality for free.

**Comment by imxieyi** [comment]
> Just need to make sure that whatever is changed, the default iOS stuff should still work like autocomplete and long press to insert accented characters, etc.

Only arrow keys on physical keyboards are blocked by raw handler. All other events should pass through and behave as usual now.

File: osu.Framework.iOS/GameUIApplication.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        public event KeyHandler KeyEvent;
+
+        // https://github.com/xamarin/xamarin-macios/blob/master/src/ObjCRuntime/Messaging.iOS.cs
+        [DllImport("/usr/lib/libobjc.dylib", EntryPoint = "objc_msgSendSuper")]
+        private static extern void send_super(IntPtr receiver, IntPtr selector, IntPtr arg1);
+
+        int lastEventFlags;
+
+        /// <summary>
+        /// Is this required?
```
Comment (peppy) [inline osu.Framework.iOS/GameUIApplication.cs]: need validation of this

Comment (imxieyi) [reply osu.Framework.iOS/GameUIApplication.cs]: As I mentioned before, the arrow key events will conflict with existing `IOSKeyboardHandler`, resulting in duplicated inputs from arrow keys. The actual behavior of this is that in map selection view, each down key will move the cursor by two rows.

Current solution is to only handle arrow keys in `IOSRawKeyboardHandler`. So `blockKeys` is needed to decide which keys are arrow keys.

Another solution is to let the events pass, then `IOSKeyboardHandler` should decide when to fire arrow key events. I don't think it's a good idea though.

Comment (peppy) [reply osu.Framework.iOS/GameUIApplication.cs]: Hmm, the arrow keys don't actually work for me with `IOSKeyboardHandler` (iPad Pro + smart keyboard). Bluetooth do already work though.

Comment (imxieyi) [reply osu.Framework.iOS/GameUIApplication.cs]: > the arrow keys don't actually work for me with `IOSKeyboardHandler`
`IOSKeyboardHandler` only recognizes keys while text input is active, e.g. `type to search` text field is active in song selection scene.

Comment (peppy) [reply osu.Framework.iOS/GameUIApplication.cs]: Okay, never mind. It works for the smart keyboard too under those conditions.

**Comment by Gentoli** [comment]
Does physical keyboard still work on iOS as of osu@2019.418.0 & osu.framework@2019.415.0?
A Magic Keyboard 1 connected to an iPhone 7 does not do anything except for text input.

**Comment by peppy** [comment]
This change never worked for me, FWIW.

**Comment by Gentoli** [comment]
Is this worth looking into? 
It seems updating graphics API will not fix the keyboard input issue(#2119).

**Comment by imxieyi** [comment]
> Does physical keyboard still work on iOS as of osu@2019.418.0 & osu.framework@2019.415.0?
> A Magic Keyboard 1 connected to an iPhone 7 does not do anything except for text input.

It should work if you change [Application.cs#L12](https://github.com/ppy/osu/blob/master/osu.iOS/Application.cs#L12) in osu into:

```c#
UIApplication.Main(args, "GameUIApplication", "AppDelegate");
```

**Pull request merged**